### PR TITLE
[Model] Add CAT (Compress and Attend Transformer)

### PR DESCRIPTION
**Pull request by bicycleman15**
Hey all, excited for my first PR at FLA 🙂 

## Summary

This PR adds Compress & Attend Transformer (CAT) to the FLA repo using existing vanilla attention components.

CAT is simple architecture that provides **inference-cost controllability at test-time**. 

On a high level, CAT decodes chunks of tokens by attending to compressed chunks of the sequence so far. Both compression and decoding can use any existing sequence mixer (we use vanilla attention for simplicity). Decoding from the compressed sequence yields compute and memory savings, with chunk size setting the operating point on the quality-cost trade-off. **Importantly, training CAT across multiple chunk sizes at once unlocks test-time control of this trade-off without any retraining, all in a single model.**

<div align="center">
  image
</div>

See the paper and upstream repo for details:

- Paper (will be updated over the weekend on arXiv): [Controllably Efficient Language Models](https://arxiv.org/abs/2511.05313)
- Reference implementation: [rajesh-lab/cat-transformer](https://github.com/rajesh-lab/cat-transformer) (esp. `cat_transformer_adaptive.py`)

Concretely, this PR adds:

1. **`CATConfig`** — decoder + compressor dims/heads/layers, `max_chunk_size`, power-of-two training schedule (`train_chunk_size_schedule`, `min_train_chunk_size`), fused-norm / SwiGLU / CE flags, etc. using existing components from FLA
2. **`CATCompressor` / `CATModel` / `CATForCausalLM`** — chunked compression, interleaved decoder sequence, resetting per-chunk RoPE (as in the upstream adaptive CAT), FlexAttention + `create_block_mask` for the CAT mask, and optional `chunk_size` override on `forward` so that inference can be done with any chunk size at test-time, without any retraining.
3. **Training schedule** — when `chunk_size` is omitted in train mode, chunk size rotates over powers of two in `[min_train_chunk_size, max_chunk_size]`; eval defaults to `max_chunk_size`.
4. **Dynamo cache limits** — raise `torch._dynamo` cache/recompile limits so cycling FlexAttention shapes stay compiled instead of falling back after the default recompile cap.
5. **AutoConfig / AutoModel / AutoModelForCausalLM registration** for `model_type="cat"`.

**Out of scope / known limits (this PR):** no KV-cache generation path yet; no variable-length `cu_seqlens` / packed batches. `CATConfig` is listed in `MODELING_UNSUPPORTED_VARLEN` and `GENERATION_UNSUPPORTED` (same pattern as other models that only support the dense fixed-length path). Full-prefix generation is smoke-tested without cache. For decoding benchmarking, upstream [cat-transformer](https://github.com/rajesh-lab/cat-transformer) repo provides those implementations. This FLA port is mainly for accessibility and "quick to try" for the model architecture community that uses FLA (like me 🙂).

## Test Plan

* `tests/models/test_modeling_cat.py`:
  * forward / backward and labeled loss
  * explicit adaptive `chunk_size` values
  * training schedule advances power-of-two chunk sizes; explicit `chunk_size` does not advance it
  * eval defaults to `max_chunk_size`
  * bucket padding / trim, all-ones attention mask, Flame-style `position_ids` + empty `cu_seqlens`
  * rejects true varlen `cu_seqlens` and padding masks
  * `chunk_size` validation
  * AutoModel registration + save/load metadata
  * full-prefix generation smoke test
* `tests/models/test_modeling_utils.py`: `CATConfig` added to the unsupported-varlen / unsupported-generation lists so shared modeling harnesses skip those paths.
* Suggested local run:

```bash
pytest tests/models/test_modeling_cat.py -q
```

## Breaking changes

None. This PR only adds new files and registrations; no existing APIs change.

## Checklist

- [x] I have read [CONTRIBUTING.md](../CONTRIBUTING.md) and follow its conventions (code style, docstrings, commit prefixes).
- [x] I have read [AGENTS.md](../AGENTS.md) and, where my change matches its scope, the relevant skill under [.agents/skills](../.agents/skills).
- [x] Dependent tests pass locally or in CI, and new behavior is covered by tests where applicable.
- [x] Kernel changes include same-hardware before/after benchmark numbers, dense + varlen where applicable (N/A: no kernel code changed).
- [ ] This PR is minor/cosmetic-only (typo, formatting, style-only tweaks).

cc @sustcsonglin @yzhangcs @zhiyuan1i

### CODE REVIEW

**Comment by zhiyuan1i** [comment]
Overall this looks good to merge — the FlexAttention path matches the upstream CAT reference, and the explicit raise for varlen/padding is appreciated. Two things I'd like to see fixed (follow-up PRs are fine):

1. `modeling_cat.py:40-52` raises `torch._dynamo` cache/recompile limits globally at import time, so importing `fla` mutates the whole process. Please scope this to the CAT compile path (or document why it must be global).
2. The 14 tests are all shape-level asserts. The mask construction and the output rearrange are the two spots that can be wrong yet still train fine — please add a numerical anchor there (padding-invariance on the original-length prefix, or a flex-vs-SDPA parity check).

Minor nits while at it: the default fused path (fuse_norm/swiglu/CE all on) has no test coverage; `output_hidden_states` returns a 1-tuple regardless of the flag, which breaks the HF contract; and `_tied_weights_keys` uses the dict form while the rest of the repo uses the list form.

**Review by zhiyuan1i — APPROVED** [review]

**Comment by bicycleman15** [comment]
heyy @zhiyuan1i 

thanks a lot for the feedback! i'll incorporate the fixes in new PRs? 

and as for this PR, i will update the README and then merge? let me know if that's okay 🙂

**Comment by bicycleman15** [comment]
hey @zhiyuan1i,

I actually went ahead and addressed all your concerns:

- Dynamo cache/recompile limits to CAT model only
- padding-invariance and exact output-rearrangement reference tests
- default fused norm/SwiGLU/CE forward-backward coverage
- fixed `output_hidden_states` to follow the HF per-layer contract
- made tied-weight metadata compatible with both Transformers 4 and 5

can you please see if this is good to merge? thanks a lot for your time!

**Review by zhiyuan1i — APPROVED** [review]
